### PR TITLE
Feature/associate categories

### LINKED ISSUE #770: Allow user to associate to categories to their product
**Is your feature request related to a problem? Please describe.**
With the implementation of #760 categories in the community we need a way to associate a product to one or more categories. 

**Describe the solution you'd like**
Modify the edit form to support categories, selecting as many as the user wants. We can follow the category screen tree-like design

When selecting a child, it should also select its parents. Although, the user can unselect freely

**Describe alternatives you've considered**
N/A

**Additional context**
The reference of the old category model: https://www.figma.com/file/OSqvmNO3HSzRJSrp9Of865/New-Shop?node-id=2624%3A27037

### PR DESCRIPTION
**Pull request by NeoVier**
## What issue does this PR close
Closes #770 

## Changes Proposed ( a list of new changes introduced by this PR)
- Add a new step on the product editing form to associate categories to a product
- Add a new button to share an offer. If you're the admin of the community, you can also click on it to edit the categories of an offer (if the community has no categories, the admin should only see the share button)

## How to test ( a list of instructions on how to test this PR)
- Associate categories to a product
- We don't have a way to filter them yet, so you need to go back to the form and see if the categories remain checked
- Test it on a community that has no categories (the categories step shouldn't show up) and make sure it still works as usual
- Log in as a community admin and make sure you can edit other people's categories
- As a community admin on a community that has no categories, make sure the option to edit categories from people's offers doesn't show up
- Test the possibility of creating offers with category

### CODE REVIEW

**Comment by NeoVier** [comment]
We're still having [some issues](https://cambiatus.slack.com/archives/CA83HJAAD/p1659651179331109) that have to be addressed on the backend, but other than that, everything should be working already

**Comment by lucca65** [comment]
Bro, do you think we can hide the category editing option if the community has no categories?

**Comment by NeoVier** [comment]
Done! ⚡

**Review by lucca65 — CHANGES_REQUESTED** [review]
Its not building for some reason, but I've also found another bug.. On Mizu we have a three layer depth in our categories tree: mens clothing → Shoes → two sub categories

The two subs are not appearing for some reason.

**Comment by NeoVier** [comment]
Build is fixed ✅ there were some problems with dependencies

Sub categories should show up now, I had some problems with recursion 🤓

**Review by lucca65 — DISMISSED** [review]
Amazing!! Almost done with categories!! 🔋

**Comment by lucca65** [comment]
@NeoVier acabei de me lembrar da questão da edição pelos Admins para as categorias. Não me lembro se discutimos isso antes, mas temos alguns elementos de tela para trabalhar isso como é visto aqui: https://www.figma.com/file/OSqvmNO3HSzRJSrp9Of865/New-Shop?node-id=2722%3A47921

**Comment by NeoVier** [comment]
Lembro de falarmos sobre isso, mas não lembro direito o que decidimos 😅. De qualquer modo, vou trabalhar nisso 🏃‍♂️

**Comment by NeoVier** [comment]
Should be done @lucca65!

The only problem now is this issue on the backend: https://github.com/cambiatus/backend/issues/277

**Comment by lucca65** [comment]
I'm getting that weird error of "error loading profile" but I can't find any faulty requests on the inspector 🤔

**Comment by lucca65** [comment]
maybe we have to sync schemas?

**Comment by NeoVier** [comment]
Seems like you had a community with logo = null. This was probably caused by https://github.com/cambiatus/backend/pull/270. On staging there was a single community that had a null logo (`0,ROTE`). I manually fixed it, but it might be worth to keep an eye out for stuff like this @MatheusBuss.

You should be able to log in as usual now @lucca65

**Comment by lucca65** [comment]
Oh i see. That's weird i thought we already did the migration that fix this in staging. let us know Matheus!

**Comment by lucca65** [comment]
Bro i'm looking from my phone while taking a boring trip home and it looks like i can acesa the edit form as an admin by clicking on the steps. Maybe we can hide the step navigator in this particular case?

IDK this makes me feel like we are having a bunch of edge cases in this feature, maybe it's too much complications, what do you think about this restriction?

**Comment by NeoVier** [comment]
The steps should already be hidden from admins 🤔 

This is what I see when I go to edit categories from someone else's product as an admin:

image

Are you sure you aren't editing your own offer? Maybe you got bit by cambiatus/backend#277 like me. I was very confused when it happened 😅.

Maybe you edited the categories from a product that wasn't yours, and it is yours now, so that's why you see the steps?

**Comment by NeoVier** [comment]
I just deployed https://github.com/cambiatus/backend/pull/278 to staging to avoid this bug

**Comment by MatheusBuss** [comment]
> Oh i see. That's weird i thought we already did the migration that fix this in staging. let us know Matheus!

The migration is already done, but it only converts empty strings to `NULL`.

I ran into a problem where some communities descriptions were empty strings but this field has the `NON_NULL` constraint, which means my script could not change those values from `''` to `NULL`.

It could happen that a field should not have `NULL`'s in it, but doensn't have the constraint enforcing this. Which could cause an error. This cou,ld be the case with the `logo` field for communities, which doesn't have the `NOT_NULL` constraint.

Perhaps we should make a small migration to change this, what do you think @NeoVier and @lucca65?

It's also worth mentioning to keep an eye out for this behavior in prod, even though I think it's very unlikely for this type of thing to happen.

**Comment by NeoVier** [comment]
Yeah, the problem happened because the *graphql schema* said that the field couldn't be null, so I guess the graphql and the sql schemas are kind of out of sync.

+1 on making a migration to make sure this doesn't happen anymore

**Review by lucca65 — APPROVED** [review]
damn, yeah bro I could have seen this one coming 😅. nice and makes perfect sense! 

Looks awesome I can't wait to finally move on from categories, or at least to actually deploy it to prod!!

File: elm-book/elm.json [hunk 1]
```diff
@@ -37,7 +37,8 @@
             "pablohirafuji/elm-syntax-highlight": "3.4.1",
```
Comment (lucca65) [inline elm-book/elm.json:37]: bro out of curiosity we still need this package? totally unrelated with the PR but i'm just curious

Comment (NeoVier) [reply elm-book/elm.json:37]: I was really confused on why we had that, until I realized it's in elm-book 🤣

We only need it as an indirect dependency for elm-book to work. I thought we were already running elm-review on elm-book, but seems like that's not the case.

I could fix it on this PR, but I ran elm-review on elm-book and uh...
image

I think it's better to have another PR to deal with it 😅 I created #809 for that

**Comment by juramos-2020** [comment]
OI @NeoVier , por favor, você me avisa quando for para testar?

**Comment by juramos-2020** [comment]
OI @NeoVier 

segue alguns feedbacks:

- Tentei criar uma nova categoria mas apareceu o feedback de erro: "Algo de errado aconteceu ao criar a categoria"

- Quando clico, em uma primeira vez, em "Editar categorias dessa oferta" abre uma página assim sem os indicadores de Steps:
image


Quando clico, em uma segunda vez, em "Editar categorias dessa oferta" da mesma oferta, já aparece os indicadores de steps conforme [design](https://www.figma.com/file/OSqvmNO3HSzRJSrp9Of865/New-Shop?node-id=2735%3A46637)
https://www.figma.com/file/OSqvmNO3HSzRJSrp9Of865/New-Shop?node-id=2735%3A46637


Acho que esse indicador deveria aparecer desde a primeira vez que clicamos em "Editar categorias dessa oferta".

**Comment by juramos-2020** [comment]
Segue os demais itens:

botão para compartilhar uma oferta ✅

Se você for o administrador da comunidade, também poderá clicar nela para editar as categorias de uma oferta (se a comunidade não tiver categorias, o administrador deverá ver apenas o botão de compartilhamento) ✅

Se não for admin da comunidade, a oferta deve apresentar apenas o botão de compártilhar ✅

Associar um produto a uma categoria ✅

Ainda não temos como filtrá-los, então você precisa voltar ao formulário e ver se as categorias continuam marcadas ✅

Teste-o em uma comunidade que não tenha categorias (a etapa de categorias não deve aparecer) e verifique se ainda funciona normalmente ✅


Faça login como administrador da comunidade e verifique se você pode editar as categorias de outras pessoas ✅

Como administrador de uma comunidade que não tem categorias, certifique-se de que a opção de editar categorias das ofertas das pessoas não apareça ✅


Ainda vou analisar o mobile para ver se teve alguma quebra de página...

**Comment by juramos-2020** [comment]
Uma dúvida @NeoVier @lucca65 [esse cenário de criação de uma oferta com categorias](https://www.figma.com/file/OSqvmNO3HSzRJSrp9Of865/New-Shop?node-id=1672%3A6427) vai entrar na entrega do item 2 ou do item 3 ? 👇

1. Liberação para admin: criar, organizar categorias
2. Edição de oferta pelo admin e associação de categorias
3. Melhorias na vitrine: exibição das categorias como filtros

**Comment by NeoVier** [comment]
> Acho que esses indicadores deveriam aparecer desde a primeira vez que clicamos em "Editar categorias dessa oferta".

Discutimos isso [aqui no figma](https://www.figma.com/file/OSqvmNO3HSzRJSrp9Of865?node-id=2735:46637#211750226), e decidimos não mostrar os steps para o admin, já que ele não pode editar os outros campos da oferta. 

Os steps apareceram na segunda vez por conta de um bug que fazia com que a oferta se tornasse sua ao editar as categorias como admin. Já temos a correção, mas o backend de staging estava em outra versão. Mudei a versão do backend de staging, então não deve mais acontecer.

---

> Uma dúvida @NeoVier @lucca65 [esse cenário de criação de uma oferta com categorias](https://www.figma.com/file/OSqvmNO3HSzRJSrp9Of865/New-Shop?node-id=1672%3A6427) vai entrar na entrega do item 2 ou do item 3 ?

Não entendi @juramos-2020. Esse link manda no cenário de recomendação/boosting de ofertas.

O item 1 já foi feito (e já fizemos deploy em produção). Nesse PR já adicionamos o item 2 e a possibilidade de criar ofertas com categorias. O item 3 vem em #806 

---

> Tentei criar uma nova categoria mas apareceu o feedback de erro: "Algo de errado aconteceu ao criar a categoria"

Vou verificar 🏃‍♂️

**Comment by juramos-2020** [comment]
> Não entendi @juramos-2020. Esse link manda no cenário de recomendação/boosting de ofertas.

Esse link aqui:
https://www.figma.com/file/OSqvmNO3HSzRJSrp9Of865/New-Shop?node-id=1672%3A6427

**Comment by juramos-2020** [comment]
> O item 1 já foi feito (e já fizemos deploy em produção). Nesse PR já adicionamos o item 2 e a possibilidade de criar ofertas com categorias. 


Nesse PR já é para testar a possibilidade de criar ofertas com categoria?

**Comment by NeoVier** [comment]
Sim 😁

**Comment by juramos-2020** [comment]
> Discutimos isso [aqui no figma](https://www.figma.com/file/OSqvmNO3HSzRJSrp9Of865?node-id=2735:46637#211750226), e decidimos não mostrar os steps para o admin, já que ele não pode editar os outros campos da oferta.
Os steps apareceram na segunda vez por conta de um bug que fazia com que a oferta se tornasse sua ao editar as categorias como admin. Já temos a correção, mas o backend de staging estava em outra versão. Mudei a versão do backend de staging, então não deve mais acontecer.

O meu comentário no figma era sobre o botão de "continue"
`Você acha que poderia abrir somente essa tela das categorias e em vez de ter o botão de "CONTINUE" ter o botão de "SALVAR"?`

Mas tranquilo, não vejo problema de não termos esse elemento visual visto que o admin não vai precisa passar pelos outos steps. Mas é bom saber para atualizar o Figma 😊

**Comment by juramos-2020** [comment]
@NeoVier 

então no caso do admin editar uma oferta de outra pessoa, os steps não aparecem, correto?
E no caso do admin editar uma oferta dele, os steps aparecem?

**Comment by NeoVier** [comment]
Isso mesmo, os steps aparecem para o dono do produto, sendo admin ou não

**Comment by juramos-2020** [comment]
Valeu @NeoVier 

então, recapitulando, seguem todos os cenarios que testei na versão desktop. Todos funcionam normalmente com excessão de não conseguir criar uma categoria.

botão para compartilhar uma oferta ✅

Se você for o administrador da comunidade, também poderá clicar nela para editar as categorias de uma oferta (se a comunidade não tiver categorias, o administrador deverá ver apenas o botão de compartilhamento) ✅

Se não for admin da comunidade, a oferta deve apresentar apenas o botão de compártilhar ✅

Associar um produto a uma categoria ✅

Ainda não temos como filtrá-los, então você precisa voltar ao formulário e ver se as categorias continuam marcadas ✅

Teste-o em uma comunidade que não tenha categorias (a etapa de categorias não deve aparecer) e verifique se ainda funciona normalmente ✅

Faça login como administrador da comunidade e verifique se você pode editar as categorias de outras pessoas ✅

Como administrador de uma comunidade que não tem categorias, certifique-se de que a opção de editar categorias das ofertas das pessoas não apareça ✅

Test the possibility of creating offers with category ✅

Admin editando uma oferta que não é dele, através de "Editar categorias dessa oferta",  não deve aparecer os steps ✅

Admin editando uma oferta que é dele, através de "Editar categorias dessa oferta" deve aparecer os step ✅

___________________________________

Vou fazer o teste na versão na versão Mobile 😎

**Comment by juramos-2020** [comment]
@NeoVier segue feedback na versão mobile:


- ⚠️ Na versão mobile também apresenta erro ao criar uma nova categpriabotão para 

- compartilhar uma oferta ✅

- ⚠️ Sobre a forma como o produto vai ser compartilhado, o Matheus ainda está trabalhando nisso correto? 
Por favor, você pode me atualizar como isso deveria aparecer?
Lembro que falamos disso em outras issues e fizemos [essa proposta](https://www.figma.com/file/OSqvmNO3HSzRJSrp9Of865/New-Shop?node-id=2665%3A44507). Mas não to lembrando em que pé está, se já foi finalizado ou não. 
Para documentar segue como está o compartilhamento:
share



- Se não for admin da comunidade, a oferta deve apresentar apenas o botão de compártilhar ✅

- Associar um produto a uma categoria  ✅

- Ainda não temos como filtrá-los, então você precisa voltar ao formulário e ver se as categorias continuam marcadas ✅

- Teste-o em uma comunidade que não tenha categorias (a etapa de categorias não deve aparecer) e verifique se ainda funciona normalmente ✅

- Se você for o administrador da comunidade, também poderá clicar nela para editar as categorias de uma oferta (se a comunidade não tiver categorias, o administrador deverá ver apenas o botão de compartilhamento) ✅ 

- ⚠️Faça login como administrador da comunidade e verifique se você pode editar as categorias de outras pessoas -  Não estou conseguindo editar, aparece o feedback de erro: "Ops Algo de errado aconteceu" 

- Como administrador de uma comunidade que não tem categorias, certifique-se de que a opção de editar categorias das ofertas das pessoas não apareça ✅

- Test the possibility of creating offers with category ✅

- Admin editando uma oferta que não é dele, através de "Editar categorias dessa oferta",  não deve aparecer os steps ✅

- Admin editando uma oferta que é dele, através de "Editar categorias dessa oferta" deve aparecer os step ✅

**Comment by MatheusBuss** [comment]
> * warning Sobre a forma como o produto vai ser compartilhado, o Matheus ainda está trabalhando nisso correto?
>       Por favor, você pode me atualizar como isso deveria aparecer?

Não foi resolvida essa issue ainda. Como é algo que vai ter uma complexidade razoável e não tem um impacto tão grande deixamos um pouco de lado.

Dito isso. Grande parte das imagens que estão nas comunidades devem funcionar normalmente. Não tenho certeza o que aconteceu com o preview que está no print. Fui testar e tanto no Slack quanto no Telegram há uma imagem, mas não a mesma que está listada no produto. Talvez seja algo a ver com o netlify. Alguma ideia @NeoVier?

image

**Comment by NeoVier** [comment]
Acho que é por conta do Netlify mesmo. O link aponta para os servidores deles, e não os nossos, então não conseguimos interceptar a request pra servir o HTML gerado pelo backend que mostraria o preview bonitinho

**Comment by NeoVier** [comment]
Fiz o deploy do frontend no staging, então podem compartilhar com o link `https://{nomeDaComunidade}.staging.cambiatus.io/shop/{idDoProduto}` e deve funcionar normalmente. Um exemplo no slack:

image

**Comment by NeoVier** [comment]
> Faça login como administrador da comunidade e verifique se você pode editar as categorias de outras pessoas - Não estou conseguindo editar, aparece o feedback de erro: "Ops Algo de errado aconteceu"

Não consegui reproduzir... Qual é a mensagem de erro exatamente, e quando ela aparece?

**Comment by juramos-2020** [comment]
> @juramos-2020 não foi resolvida essa issue ainda. Como é algo que vai ter uma complexidade razoável e não tem um impacto tão grande deixamos um pouco de lado.

Tranquilo @MatheusBuss 

> Fiz o deploy do frontend no staging, então podem compartilhar com o link https://{nomeDaComunidade}.staging.cambiatus.io/shop/{idDoProduto} e deve funcionar normalmente.

Pra mim ainda aparece dessa forma 👇.  Mas assim, o compartilhamento está funcinando pois está compartilhando. Só estava na dúvida sobre a aparencia desse compartilhamento. Sabia que o @MatheusBuss estava trabalhando nisso só não tinha certeza se já tinha sido concluído ou não. Então fica apenas o registro de como tá o compartilhamento para documentar mesmo.
Share 01

**Comment by NeoVier** [comment]
É que o link tá errado Ju 😅. Pro preview do link funcionar tem que ser pelo link do frontend de staging (e não pelo gerado aqui no Github). No caso desse produto é https://nomad.staging.cambiatus.io/shop/74

**Comment by juramos-2020** [comment]
> Não consegui reproduzir... Qual é a mensagem de erro exatamente, e quando ela aparece?

@NeoVier Tentei reproduzir agora para filmar mas não aconteceu, acho que de alguma forma foi resolvido, que bom rsrs 😎

**Comment by NeoVier** [comment]
Bom, o único problema então é na hora de criar categoria... Como isso não está relacionado a esse PR, vou fazer o merge desse e vemos isso em outro PR. Já sinalizei o problema aqui no Slack: https://cambiatus.slack.com/archives/CA83HJAAD/p1660648202151329

**Comment by juramos-2020** [comment]
> É que o link tá errado Ju 😅. Pro preview do link funcionar tem que ser pelo link do frontend de staging (e não pelo gerado aqui no Github). No caso desse produto é https://nomad.staging.cambiatus.io/shop/74

Ahhhh bom, assim sim hahaha 😂


Arrasaram meninos ficou show!!!!

StandingOvationGIF

**Comment by juramos-2020** [comment]
> Bom, o único problema então é na hora de criar categoria... Como isso não está relacionado a esse PR, vou fazer o merge desse e vemos isso em outro PR. Já sinalizei o problema aqui no Slack: https://cambiatus.slack.com/archives/CA83HJAAD/p1660648202151329

Combinado!

**Comment by lucca65** [comment]
Animal pessoal, vamos ajustar essas arestas aí, mas to contente que nessa semana sai nosso querido update final de categorias!!

**Pull request merged**